### DISCUSSION TITLE
Learning Python: Exploring a Maze Generator with Tkinter

### DISCUSSION THREAD

**Discussion by JamesCodingSpace** [JamesCodingSpace/Main#14] [General]
Hey fellow Python enthusiasts! Today, I want to share a simple yet exciting project I've been working on to enhance my Python skills: a [Maze Generator](https://github.com/JamesCodingSpace/Main/blob/codespace-urban-carnival-4jjvwx5vq6p6cgg/notebooks/python/labyrith.py) using the Tkinter library. Tkinter is a fantastic library for creating graphical user interfaces, making it a perfect choice for this project.

### **Setting Up the GUI**
To begin with, let's dive into the code. The first thing I noticed was the use of the Tkinter library for creating the graphical user interface (GUI). The MazeGeneratorGUI class serves as the backbone for our application.

```
import tkinter as tk
import random

class MazeGeneratorGUI:
    def __init__(self, master):
        # ... (Other initialization code)

        # Labyrinthgröße (Maze size) Label and Entry
        self.size_label = tk.Label(master, text="Labyrinthgröße:")
        self.size_label.pack()

        self.size_entry = tk.Entry(master)
        self.size_entry.pack()

        # Schwierigkeitsstufe (Difficulty Level) Label and OptionMenu
        self.difficulty_label = tk.Label(master, text="Schwierigkeitsstufe:")
        self.difficulty_label.pack()

        self.difficulty_var = tk.StringVar()
        self.difficulty_var.set("easy")

        difficulty_options = ["easy", "medium", "hard"]
        self.difficulty_menu = tk.OptionMenu(master, self.difficulty_var, *difficulty_options)
        self.difficulty_menu.pack()

        # Labyrinth generieren (Generate Maze) Button
        self.generate_button = tk.Button(master, text="Labyrinth generieren", command=self.generate_maze)
        self.generate_button.pack()

        # Canvas for drawing the maze
        self.canvas = tk.Canvas(master, width=500, height=500)
        self.canvas.pack()

    # ... (Other methods)
```
In this snippet, we set up various GUI elements such as labels, entry fields, an option menu, and a canvas to draw the maze. The generate_maze method is linked to the "Labyrinth generieren" button, which triggers the maze generation process.

### **Generating the Maze**
The generate_maze_data method handles the actual maze generation. It takes the size and difficulty level as inputs, then populates a 2D list (maze) with values representing walls (1) and open paths (0).

```
def generate_maze_data(self, size, difficulty):
    if difficulty == "easy":
        wall_density = 0.1
    elif difficulty == "medium":
        wall_density = 0.3
    elif difficulty == "hard":
        wall_density = 0.5
    else:
        raise ValueError("Ungültige Schwierigkeitsstufe")

    maze = [[0] * size for _ in range(size)]

    # Set maze boundaries
    for i in range(size):
        maze[0][i] = maze[size - 1][i] = 1
        maze[i][0] = maze[i][size - 1] = 1

    # Set walls inside the maze
    for i in range(1, size - 1):
        for j in range(1, size - 1):
            if random.random() < wall_density:
                maze[i][j] = 1

    # Set entrance and exit
    maze[0][1] = maze[size - 1][size - 2] = 0

    return maze
```
This method uses the difficulty level to determine the density of walls in the maze. The generated maze is a 2D list with boundaries, walls, and designated entrance and exit points.

### **Drawing the Maze**
The draw_maze method handles the visualization of the generated maze on the canvas.

```
def draw_maze(self, maze):
    self.canvas.delete("all")

    cell_size = 500 / len(maze)

    for i in range(len(maze)):
        for j in range(len(maze[i])):
            x0, y0 = j * cell_size, i * cell_size
            x1, y1 = (j + 1) * cell_size, (i + 1) * cell_size

            if maze[i][j] == 1:
                self.canvas.create_rectangle(x0, y0, x1, y1, fill="black")
            else:
                self.canvas.create_rectangle(x0, y0, x1, y1, fill="white")
```
This method utilizes the Tkinter Canvas widget to draw rectangles based on the values in the maze data. Walls are represented by black rectangles, while open paths are represented by white rectangles.

### **Running the Application**
Finally, we use the if __name__ == "__main__": block to instantiate the Tkinter application and run the event loop.

```
if __name__ == "__main__":
    root = tk.Tk()
    app = MazeGeneratorGUI(root)
    root.mainloop()
```
This block ensures that the code is only executed when the script is run directly, not when it's imported as a module.

That's it! With this Maze Generator, I've not only explored Tkinter but also honed my skills in working with 2D arrays and randomization in Python. I hope this walkthrough helps you on your Python learning journey! Happy coding!

**Comment by PhilippR84** [comment]
WOW sehr interessant! Ein Bild zu dem Labyrinth wäre schön! Sonst sehr gut.

**Comment by nklshpf** [comment]
Very interesting! It would be very cool if you could show us a picture of the final outcome when running the code. Maybe in the next discussion. But still impressive. Keep it up!
